src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1762,31 +1901,107 @@ public sealed class One : Tests
+                    if (test.Input.Source is BoundDagPropertyEvaluation { IsLengthOrCount: true } e)
+                    {
+                        if (e.Syntax.IsKind(SyntaxKind.ListPattern))
+                        {
+                            // Normally this kind of test is never created, we do this here because we're rewriting nested tests
+                            // to check the top-level length instead. If we never create this test, the length temp gets removed.
+                            if (test is BoundDagRelationalTest t)
+                            {
+                                Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                Debug.Assert(t.Relation == BinaryOperatorKind.GreaterThanOrEqual);
+                                Debug.Assert(t.Value.Int32Value >= 0);
+                                if (t.Value.Int32Value == 0)
+                                    return True.Instance;
+                            }
+                        }
+                        if (TryGetTopLevelLengthTemp(e) is (BoundDagTemp lengthTemp, int offset))
+                        {
+                            // If this is a nested length test in a slice subpattern, update the matched
+                            // value based on the number of additional elements in the containing list.
+                            switch (test)
+                            {
+                                case BoundDagValueTest t when !t.Value.IsBad:
+                                    Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                    return new One(new BoundDagValueTest(t.Syntax,
+                                        ConstantValue.Create(safeAdd(t.Value.Int32Value, offset)), lengthTemp));
+                                case BoundDagRelationalTest t when !t.Value.IsBad:
+                                    Debug.Assert(t.Value.Discriminator == ConstantValueTypeDiscriminator.Int32);
+                                    return new One(new BoundDagRelationalTest(t.Syntax, t.OperatorKind,
+                                        ConstantValue.Create(safeAdd(t.Value.Int32Value, offset)), lengthTemp));
+                            }
+                        }
+                    }
+
+                    return this;
+
+                    static int safeAdd(int value, int offset)
+                    {
+                        Debug.Assert(value >= 0); // Tests with negative values should never be created.
+                        Debug.Assert(offset >= 0); // The number of elements in a list is always positive.
+                        // We don't expect offset to be very large, but the value could
+                        // be anything given that it comes from a constant in the source.
+                        return offset > (int.MaxValue - value) ? int.MaxValue : value + offset;
```

Comment:
Will there be some error if this condition is true?

Reply:
No, consider: `[_, ..{ Length: <= Max }]` the nested length would still check for Max, not one more, clamped within the valid range.

Reply:
🚧  Actually, it makes sense to say this is not going to match (the nested Lengthin can never be equal to MaxValue)